src/momento/aio/_scs_data_client.py

```diff
@@ -60,27 +61,173 @@ async def set(
                 metadata=_make_metadata(cache_name),
                 timeout=self._default_deadline_seconds,
             )
-            _momento_logger.debug(f"Set succeeded for key: {key}")
-            return cache_sdk_resp.CacheSetResponse(response, set_request.cache_body)
+            _momento_logger.debug(f"Set succeeded for key: {str(key)}")
+            return cache_sdk_ops.CacheSetResponse(
+                response, set_request.cache_key, set_request.cache_body
+            )
+        except Exception as e:
+            _momento_logger.debug(f"Set failed for {str(key)} with response: {e}")
+            raise _cache_service_errors_converter.convert(e)
+
+    async def multi_set(
+        self,
+        cache_name: str,
+        set_operations: Union[
+            List[cache_sdk_ops.CacheMultiSetOperation],
+            List[cache_sdk_ops.CacheMultiSetFailureResponse],
+        ],
+    ) -> cache_sdk_ops.CacheMultiSetResponse:
+
+        _validate_multi_op_list(set_operations)
+        _validate_cache_name(cache_name)
+
+        # Will hold which tasks have succeeded or failed to return to caller
+        failed_ops: List[cache_sdk_ops.CacheMultiSetFailureResponse] = []
+        successful_ops: List[cache_sdk_ops.CacheSetResponse] = []
+
+        try:
+            request_promises = list()
+            for op in set_operations:
+                item_ttl_seconds = (
+                    self._default_ttlSeconds
+                    if op.ttl_seconds is None
+                    else op.ttl_seconds
+                )
+                _validate_ttl(item_ttl_seconds)
+                set_request = _SetRequest()
+                set_request.cache_key = _as_bytes(op.key, "Unsupported type for key: ")
+                set_request.cache_body = _as_bytes(
+                    op.value, "Unsupported type for value: "
+                )
+                set_request.ttl_milliseconds = item_ttl_seconds * 1000
+
+                request_promises.append(
+                    self._grpc_manager.async_stub().Set(
+                        set_request,
+                        metadata=_make_metadata(cache_name),
+                        timeout=self._default_deadline_seconds,
+                    )
+                )
+
+            results = await asyncio.gather(
+                *request_promises,
+                return_exceptions=True,
+                # When set to True exceptions are treated the same as successful results, and aggregated in the
+                # result list. We want to try and make sure all requests have chance to finish.
+            )
+
+            for op, result in zip(set_operations, results):  # type: ignore[assignment]
+                key = _as_bytes(op.key, "Unsupported type for key: ")
```

Comment:
Might be better to either hold onto the request or send back the same key object rather than re-converting and doubling the key and value memory, object/gc and cpu byte walking overhead. The way this is structured it's charging the user 2 conversions for every successful operation right?